pkg/cloud/externalconn/impl_registry.go

```diff
@@ -46,25 +50,27 @@ type schemeRegistration struct {
 	// connectionParserFactory is the function that will be called
 	// to create an ExternalConnection from a URI of a given scheme.
 	connectionParserFactory
-	// validations contains all registered validations specific to this scheme,
-	// indexed by name or the defaultValidation magic constant.
-	validations map[string]ValidationFn
+	// validationsWithCloud contains all registered validations for the cloud service providers
+	// specific to this scheme, indexed by name or the defaultValidation magic constant.
+	validationsWithCloud map[string]ValidationFnWithCloud
+	// validationsWithSinkURI contains the validations from the changefeedccl
+	// side. We want to distinguish it from the one from the cloud provider side.
+	validationsWithSinkURI map[string]ValidationFnWithSinkURI
 }
 
 func (s *schemeRegistration) parseAndValidateURI(
-	ctx context.Context,
-	execCfg interface{},
-	user username.SQLUsername,
-	uri *url.URL,
-	requestedValidations ...string,
+	ctx context.Context, env ExternalConnEnv, uri *url.URL, requestedValidations ...string,
 ) (ExternalConnection, error) {
-	conn, err := s.connectionParserFactory(ctx, execCfg, user, uri)
+	conn, err := s.connectionParserFactory(ctx, env, uri)
 	if err != nil {
 		return nil, err
 	}
 	for _, v := range requestedValidations {
-		if validationFn, ok := s.validations[v]; ok {
-			err = errors.CombineErrors(err, validationFn(ctx, execCfg, user, uri.String()))
+		if validationFn, ok := s.validationsWithCloud[v]; ok {
+			err = errors.CombineErrors(err, validationFn(ctx, cloud.KMSEnv(&env), uri.String()))
+		}
+		if validationFn, ok := s.validationsWithSinkURI[v]; ok {
+			err = errors.CombineErrors(err, validationFn(ctx, env, uri.String()))
```

Comment:
similar to my other comment I like the pattern of passing in an ExternalConnEnv to every validation method. It is then up to the validation methods to use that env to construct any validation-specific state. If we do that, I think we can revert this diff and go back to having a single validations map?